### PR TITLE
type/util: add functions mixDateAndTime and getDateFromDaynr

### PR DESCRIPTION
**Pull request by tiancaiamao**
I thought that `mixDateAndTime` could handle value overflow during parse time, so I do this. 
But later I find it still can't handle month or day contains 0.

Maybe they could be used to implement `from_day` in the future.

### CODE REVIEW

File: util/types/mytime_test.go [hunk 1]
```diff
@@ -104,3 +104,66 @@ func (s *testMyTimeSuite) TestCompareTime(c *C) {
...
+		},
+		{
+			date:   mysqlTime{2016, 12, 0, 0, 0, 0, 0},
+			time:   mysqlTime{0, 0, 0, 24, 0, 0, 0},
+			expect: mysqlTime{2016, 12, 1, 0, 0, 0, 0},
+		},
+	}
+
+	for _, t := range cases {
+		mixDateAndTime(&t.date, &t.time, false)
```
Comment (hanfei1991) [inline util/types/mytime_test.go]: add test that neg is true.

File: util/types/mytime.go [hunk 2]
```diff
@@ -276,3 +277,75 @@ func calcWeek(t *mysqlTime, wb weekBehaviour) (year int, week int) {
...
+func mixDateAndTime(date, time *mysqlTime, neg bool) {
+	if !neg && time.hour < 24 {
+		date.hour = time.hour
+		date.minute = time.minute
+		date.second = time.second
+		date.microsecond = time.microsecond
+		return
+	}
+
+	// time is negative or outside of 24 hours internal.
```
Comment (hanfei1991) [inline util/types/mytime.go]: time -> Time

File: util/types/mytime.go [hunk 3]
```diff
@@ -276,3 +277,75 @@ func calcWeek(t *mysqlTime, wb weekBehaviour) (year int, week int) {
...
+	sign := -1
+	if neg {
+		sign = 1
+	}
+	seconds, useconds, neg := calcTimeDiff(date, time, sign)
```
Comment (hanfei1991) [inline util/types/mytime.go]: What is the meaning of `useconds` ?

File: util/types/mytime.go [hunk 4]
```diff
@@ -276,3 +277,75 @@ func calcWeek(t *mysqlTime, wb weekBehaviour) (year int, week int) {
...
+		year++
+		daysInYear = calcDaysInYear(int(year))
+	}
+
+	leapDay := uint(0)
+	if daysInYear == 366 {
+		if dayOfYear > 31+28 {
+			dayOfYear--
+			if dayOfYear == 31+28 {
+				// Handle leapyears leapday
```
Comment (hanfei1991) [inline util/types/mytime.go]: add `.`

File: util/types/mytime.go [hunk 5]
```diff
@@ -276,3 +277,75 @@ func calcWeek(t *mysqlTime, wb weekBehaviour) (year int, week int) {
...
+	sign := -1
+	if neg {
+		sign = 1
+	}
+	seconds, useconds, neg := calcTimeDiff(date, time, sign)
+
+	// If we want to use this function with arbitrary dates, this code will need
+	// to cover cases when time is negative and "date < -time".
+
+	days := seconds / secondsIn24Hour
```
Comment (hanfei1991) [inline util/types/mytime.go]: swap line 301 and 302 ?

**Comment by XuHuaiyu** [comment]
Rest LGTM.

File: util/types/mytime.go (hunk 5 above)
Comment (tiancaiamao) [inline util/types/mytime.go]: Unnecessary, they are equivalent.

**Comment by hanfei1991** [comment]
LGTM

**Comment by tiancaiamao** [comment]
PTAL @XuHuaiyu

**Comment by XuHuaiyu** [comment]
LGTM, please resolve the conflicts.

**Pull request merged**